### ISSUE TITLE
Feature Request -> Export to HTML

### ISSUE THREAD

**Issue by spmeesseman** [fabiospampinato/vscode-todo-plus#185]
Great great great extension, I really see value in this, especially if I could export the marked up file to HTML to send to co workers and superiors.  Any chance for this in the future?  I'd take a stab at it myself if you think its not worth your time, as soon as I can get some time myself haha.  But thanks again for a great tool in any case

**Comment by fabiospampinato** [comment]
If you'd like to submit a PR for this I'd be happy to consider it.

It shouldn't be too difficult to implement, but I'm not sure how "cancelled" todos should be translated to HTML, since there's no concept of a "cancelled" checkbox 🤔

**Comment by pzmarzly** [comment]
Maybe [indeterminate checkboxes](https://css-tricks.com/indeterminate-checkboxes/) could be used? They display as `[-]` in modern browsers (not sure about Safari though).

**Comment by fabiospampinato** [comment]
@pzmarzly Maybe 🤷‍♂️ Maybe we could get a nice output if we styled those via CSS to make them appear more like "cancelled" todos rather than indeterminate.

**Comment by spmeesseman** [comment]
was going to take a poke at this.  i need a little help though i am not too familiar with ts, just started using it a few weeks ago, why do I get these 5 errors like this when trying to build a base branch?  I mean, its obvious, but the way you wrote that it seems to be like the compiler is supposed to handle that, checking the instance types of the variable and coding accordingly.  WHat am I missing?

`ERROR in C:\Projects\vscode-todo-plus\src\todo\document.ts
./src/todo/document.ts
[tsl] ERROR in C:\Projects\vscode-todo-plus\src\todo\document.ts(23,7)
      TS2322: Type 'string | TextDocument | TextEditor' is not assignable to type 'string'.
  Type 'TextDocument' is not assignable to type 'string'.

ERROR in C:\Projects\vscode-todo-plus\src\todo\document.ts
./src/todo/document.ts
[tsl] ERROR in C:\Projects\vscode-todo-plus\src\todo\document.ts(27,26)
      TS2361: The right-hand side of an 'in' expression must be of type 'any', an object type or a type parameter.

ERROR in C:\Projects\vscode-todo-plus\src\utils\time.ts
./src/utils/time.ts
[tsl] ERROR in C:\Projects\vscode-todo-plus\src\utils\time.ts(125,15)
      TS2339: Property 'replace' does not exist on type 'string | number'.
  Property 'replace' does not exist on type 'number'.

ERROR in C:\Projects\vscode-todo-plus\src\utils\time.ts
./src/utils/time.ts
[tsl] ERROR in C:\Projects\vscode-todo-plus\src\utils\time.ts(126,15)
      TS2339: Property 'replace' does not exist on type 'string | number | Date'.
  Property 'replace' does not exist on type 'number'.

ERROR in C:\Projects\vscode-todo-plus\src\utils\time.ts
./src/utils/time.ts
[tsl] ERROR in C:\Projects\vscode-todo-plus\src\utils\time.ts(128,33)
      TS2345: Argument of type 'string | number | Date' is not assignable to parameter of type 'string'.
  Type 'number' is not assignable to type 'string'.`

**Comment by fabiospampinato** [comment]
> why do I get these 5 errors like this when trying to build a base branch?

@spmeesseman I'm not sure what that means, did you clone the repo? How are you building the extension? Those look like some type errors, I'm not sure why I'm not seeing them (maybe you've got a different TS version, or downloaded different types for vscode).

> the way you wrote that it seems to be like the compiler is supposed to handle that

I'm not sure what this means either, they way I wrote what? What's the compiler supposed to handle?

**Comment by spmeesseman** [comment]
Yes the repo is cloned with no modification, yes they are type errors and Im not sure how its compiling for you, I am using ts 2.9.8 which is max version semantically you have defined in package.json.

If I cast, the errors go away. example, document.ts:

`if ( _.isString ( res ) ) {

      this.text = res as string;

    }`

works fine

**Comment by fabiospampinato** [comment]
Yeah but that's just like telling the type checker to shut up. I'm not sure why you're getting those errors, if I run `npm run compile` I don't get any 🤔

**Comment by spmeesseman** [comment]
yeah, weird.  It runs fine and I can debug the extension though.  Gonna try to spend some time trying to figure out how it works this weekend.  Do you have any suggestions on where to start looking to see how you color the tags and text and time calculations in the underlying todo document of the active editor?

**Comment by fabiospampinato** [comment]
> Do you have any suggestions on where to start looking to see how you color the tags and text and time calculations in the underlying todo document of the active editor?

All the supported decorations are defined [here](https://github.com/fabiospampinato/vscode-todo-plus/tree/master/src/todo/decorators), each one of those files basically maps regexes to decorations.

Most of the logic regarding individual todos is [here](https://github.com/fabiospampinato/vscode-todo-plus/blob/master/src%2Ftodo%2Fitems%2Ftodo.ts).

But if you want to implement an export to HTML feature you just have to retrieve the current editor, check that it is of a supported language, and then basically transform each line to HTML, you can walk the document via [this](https://github.com/fabiospampinato/vscode-todo-plus/blob/fb764a46040b72b8aa093977e1fc9e34c28a6e29/src/utils/ast.ts#L64) function.

**Comment by teterkin** [comment]
> It shouldn't be too difficult to implement, but I'm not sure how "cancelled" todos should be translated to HTML, since there's no concept of a "cancelled" checkbox 🤔
You can try this: ✘ Test (25-04-19 16:56)

**Comment by spmeesseman** [comment]
Thanks...
I have a little something done, when I have some more time I'll check back in with you to see what you think.  It needs some work and I have some questions but wanted to drop a note so you knew i didn;t disappear, just really busy at work
